src/wp-includes/classicpress/class-wp-compat.php

```diff
@@ -0,0 +1,310 @@
+			return;
+		}
+
+		if ( $theme_using_blocks === '1' ) {
+			// Translators: %1$s is the theme name.
+			$message = sprintf( esc_html__( '%1$s uses block-related functions and may not work correctly.' ), wp_get_theme()->get( 'Name' ) );
+		} else {
+			// Translators: %1$s is the theme name, %1$s is the parent theme name.
+			$message = sprintf( esc_html__( '%1$s parent theme (%2$s) uses block-related functions and may not work correctly.' ), wp_get_theme()->get( 'Name' ), wp_get_theme()->parent()->get( 'Name' ) );
+		}
+
+		?>
+		<div class="notice notice-alt notice-warning">
+			<p>
+			<?php
+				echo $message;
+			?>
+		</div>
+		<?php
+	}
+
+	/**
+	 * Get plugin folder name from a path.
+	 *
+	 * @param string $path
+	 * @return string
+	 */
+	private function plugin_folder( $path ) {
+		return preg_replace( '~^' . preg_quote( WP_PLUGIN_DIR ) . preg_quote( DIRECTORY_SEPARATOR ) . '([^' . preg_quote( DIRECTORY_SEPARATOR ) . ']*).*~', '$1', $path );
+	}
+
+	/**
+	 * Action hooked to after_switch_theme to remove the theme
+	 * that may not work properly.
+	 *
+	 * @return void
+	 */
+	public function delete_theme_uses_blocks() {
+		update_option( 'theme_using_blocks', '0' );
+	}
+
+
+	/**
+	 * This function have to be called from a polyfill
+	 * to map themes and plugins calling those functions.
+	 *
+	 * @return void
+	 */
+	public function using_block_function() {
+		if ( $this->blocks_compatibility_level !== 2 ) {
+			return;
+		}
+		$trace = debug_backtrace();
+		if ( strpos( $trace[1]['file'], get_stylesheet_directory() ) === 0 ) {
+			// Current theme is calling the function
+			update_option( 'theme_using_blocks', '1' );
+		} elseif ( strpos( $trace[1]['file'], get_template_directory() ) === 0 ) {
+			// Parent theme is calling the function
+			update_option( 'theme_using_blocks', '2' );
+		} else {
+			// A plugin is calling the function
+			$traces = array_column( $trace, 'file' );
+			$traces = array_map(
+				function( $path ) {
+					return $this->plugin_folder( $path );
+				},
+				$traces
+			);
+			$active = wp_get_active_and_valid_plugins();
+			$active = array_map(
+				function( $path ) {
+					return $this->plugin_folder( $path );
+				},
+				$active
+			);
+			$plugins = array_intersect( $traces, $active );
+			$plugin = array_pop( $plugins );
+			if ( $plugin === null ) {
+				// Nothing found? Bail.
+				return;
+			}
+			$plugins_using_blocks = get_option( 'plugins_using_blocks', array() );
+			if ( ! array_key_exists( plugin_basename( $plugin ), $plugins_using_blocks ) ) {
+				$plugins_using_blocks[ plugin_basename( $plugin ) ] = true;
+				update_option( 'plugins_using_blocks', $plugins_using_blocks );
+			}
+		}
+	}
+
+	/**
+	 * Polyfills that can not be defined elsewhere go here.
+	 * that may not work properly.
+	 *
+	 * @return void
+	 */
+	private function define_polyfills() {
+		// Polyfills that must not have to be defined elsewhere go there.
+		if ( ! function_exists( 'register_block_type' ) ) :
```

Comment:
I think we should aim to use `{}` rather than `: endif;` enclosures as per preferred coding standards.